### PR TITLE
fix server_test.go so it works + make code flow test pass with no secret

### PR DESCRIPTION
**Pull request by jarlandre**
1. Remove query encoding of urls for redirect uri in server_test.go to make tests successfully hit the /oauth2 callback endpoint
2. Then fix the token requests in the auth code flow tests
3. Then fix the manager to allow for no client_secret for auth code flow with pkce where code verifier is at least 43 chars. EDIT: i see there might not be need to check its length and only its existence. Its already validated properly. Merely checking that that code verifier is not empty string will be enough to ensure that the request will fail without the proper code verifier. 

- [x] Remove length check on code verifier and only check if its a non-empty string
- [ ] actually we need to differentiate between confidential and public clients, which this server library cannot, afaik. EDIT: new draft PR here https://github.com/go-oauth2/oauth2/pull/233

the pkce spec states that code verifier should be minimum 43 chars (https://github.com/go-oauth2/oauth2/pull/230#issuecomment-1396556389), should this be verified earlier in the process?

### CODE REVIEW

**Comment by jarlandre** [comment]
now im going to debug why this happens

**Comment by jarlandre** [comment]
access_token is blank string "" for both plain and 256 auth code tests. The debugging makes no sense. 

Screenshot 2023-01-18 at 21 39 24

**Comment by jarlandre** [comment]
so, any feedback on this is deeply appreciated. I dont see a simple fix for this sadly

**Comment by jarlandre** [comment]
oh and btw the change from clientinfobasic to clientinfoform isnt relevant. It fails with either.

**Comment by jarlandre** [comment]
its all because of wrongly shared code. In this case when the token endpoint is called with authorization_code grant type client_secret should not be required

Screenshot 2023-01-18 at 22 22 42

File: manage/manager.go [hunk 1]
```diff
@@ -287,7 +287,7 @@ func (m *Manager) GenerateAccessToken(ctx context.Context, gt oauth2.GrantType,
 		if !cliPass.VerifyPassword(tgr.ClientSecret) {
 			return nil, errors.ErrInvalidClient
 		}
-	} else if len(cli.GetSecret()) > 0 && tgr.ClientSecret != cli.GetSecret() {
+	} else if len(cli.GetSecret()) > 0 && tgr.ClientSecret != cli.GetSecret() && !(gt.String() == oauth2.AuthorizationCode.String() && tgr.ClientSecret == "") {
```
Comment (jarlandre) [inline manage/manager.go:290]: fix with this 

```go
!(gt.String() == oauth2.AuthorizationCode.String() && tgr.ClientSecret == "")
```

Comment (jarlandre) [reply manage/manager.go:290]: now all tests passes

Comment (jarlandre) [reply manage/manager.go:290]: so now some pro devs on this library can come up with a better fix for it 🙏

**Comment by jarlandre** [comment]
☝️ @LyricTian

File: manage/manager.go (hunk 1 above)
Comment (jarlandre) [reply manage/manager.go]: removed the .String() stuff.. was something added while i debugged or sumtin

File: manage/manager.go [hunk 2]
```diff
@@ -288,7 +288,11 @@ func (m *Manager) GenerateAccessToken(ctx context.Context, gt oauth2.GrantType,
 			return nil, errors.ErrInvalidClient
 		}
 	} else if len(cli.GetSecret()) > 0 && tgr.ClientSecret != cli.GetSecret() {
-		return nil, errors.ErrInvalidClient
+		// auth code flow doesnt require client_secret if used with PKCE and state parameter
+		// this is especially useful for mobile apps, that cant hold the secret
+		if !(gt == oauth2.AuthorizationCode && tgr.ClientSecret == "" && tgr.CodeVerifier != "") {
```
Comment (jarlandre) [inline manage/manager.go:293]: this should only affect plain or hashed PKCE flow. Not sure if its possible to differentiate between plain and hashed at this point, but if it is we should do it.

**Comment by jarlandre** [comment]
```
code_verifier — The code verifier should be a high-entropy cryptographic random string with a minimum of 43 characters and a maximum of 128 characters.
```

**Comment by jarlandre** [comment]
```
./go.test.sh
ok  	github.com/go-oauth2/oauth2/v4	0.203s	coverage: 42.9% of statements
?   	github.com/go-oauth2/oauth2/v4/errors	[no test files]
?   	github.com/go-oauth2/oauth2/v4/example/client	[no test files]
?   	github.com/go-oauth2/oauth2/v4/example/server	[no test files]
ok  	github.com/go-oauth2/oauth2/v4/generates	0.219s	coverage: 77.8% of statements
ok  	github.com/go-oauth2/oauth2/v4/manage	0.211s	coverage: 61.8% of statements
?   	github.com/go-oauth2/oauth2/v4/models	[no test files]
ok  	github.com/go-oauth2/oauth2/v4/server	0.237s	coverage: 53.6% of statements
ok  	github.com/go-oauth2/oauth2/v4/store	2.314s	coverage: 87.2% of statements
```

**Comment by DennisMuchiri** [comment]
@jarlandre why was this not merged?

**Comment by jarlandre** [comment]
because it war replaced by another PR .. see mentions

**Comment by jarlandre** [comment]
should have rebased and kept it in this PR i guess .. live and learn

**Comment by shynome** [comment]
here is a tip: how to work with pcke , you need change the ClientInfoHandler to resolve `invalid_client` error
```golang
srv.SetClientInfoHandler(server.ClientFormHandler)
```

**Pull request closed without merging**